<a href="https://colab.research.google.com/github/bitnarahyo/deep-learning-with-python-notebooks/blob/master/2_1_a_first_look_at_a_neural_network_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
 %tensorflow_version 1.x 

import keras
# 케라스 패키지 활성화
keras.__version__
# 케라스 버전 확인

TensorFlow 1.x selected.


Using TensorFlow backend.


'2.2.5'

# A first look at a neural network

This notebook contains the code samples found in Chapter 2, Section 1 of [Deep Learning with Python](https://www.manning.com/books/deep-learning-with-python?a_aid=keras&a_bid=76564dff). Note that the original text features far more content, in particular further explanations and figures: in this notebook, you will only find source code and related comments.

----

We will now take a look at a first concrete example of a neural network, which makes use of the Python library Keras to learn to classify 
hand-written digits. Unless you already have experience with Keras or similar libraries, you will not understand everything about this 
first example right away. You probably haven't even installed Keras yet. Don't worry, that is perfectly fine. In the next chapter, we will 
review each element in our example and explain them in detail. So don't worry if some steps seem arbitrary or look like magic to you! 
We've got to start somewhere.

The problem we are trying to solve here is to classify grayscale images of handwritten digits (28 pixels by 28 pixels), into their 10 
categories (0 to 9). The dataset we will use is the MNIST dataset, a classic dataset in the machine learning community, which has been 
around for almost as long as the field itself and has been very intensively studied. It's a set of 60,000 training images, plus 10,000 test 
images, assembled by the National Institute of Standards and Technology (the NIST in MNIST) in the 1980s. You can think of "solving" MNIST 
as the "Hello World" of deep learning -- it's what you do to verify that your algorithms are working as expected. As you become a machine 
learning practitioner, you will see MNIST come up over and over again, in scientific papers, blog posts, and so on.

The MNIST dataset comes pre-loaded in Keras, in the form of a set of four Numpy arrays:

In [6]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
# 케라스 내 데이터중 mnist 데이터를 가져와서 데이터 세트 구성

11493376/11490434 [==============================] - 2s 0us/step


`train_images` and `train_labels` form the "training set", the data that the model will learn from. The model will then be tested on the 
"test set", `test_images` and `test_labels`. Our images are encoded as Numpy arrays, and the labels are simply an array of digits, ranging 
from 0 to 9. There is a one-to-one correspondence between the images and the labels.

Let's have a look at the training data:

In [7]:
train_images.shape
# 학습 데이터 이미지 dimension

(60000, 28, 28)

In [0]:
len(train_labels)
# 학습데이터 라벨 개수

60000

In [8]:
train_labels
# 학습데이터 라벨 출력

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

Let's have a look at the test data:

In [9]:
test_images.shape
# 테스트데이터 이미지 dimension

(10000, 28, 28)

In [0]:
len(test_labels)
# 테스트데이터 라벨 개수

10000

In [10]:
test_labels
# 테스트데이터 라벨 출력

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

Our workflow will be as follow: first we will present our neural network with the training data, `train_images` and `train_labels`. The 
network will then learn to associate images and labels. Finally, we will ask the network to produce predictions for `test_images`, and we 
will verify if these predictions match the labels from `test_labels`.

Let's build our network -- again, remember that you aren't supposed to understand everything about this example just yet.

In [11]:
from keras import models
from keras import layers
# 케라스 패키지에서 모형과 레이어 모듈 불러들이기

network = models.Sequential()
# 레이어를 선형으로 연결하는 sequential 모형을 사용
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
# input data는 (28*28)형태로 넣어 'relu'함수를 적용해 512 형태로 나오는 Dense 레이어를 네트워크에 추가
network.add(layers.Dense(10, activation='softmax'))
# 'softmax'함수를 적용해 10 형태로 나오는 Dense 레이어를 네트워크에 추가



The core building block of neural networks is the "layer", a data-processing module which you can conceive as a "filter" for data. Some 
data comes in, and comes out in a more useful form. Precisely, layers extract _representations_ out of the data fed into them -- hopefully 
representations that are more meaningful for the problem at hand. Most of deep learning really consists of chaining together simple layers 
which will implement a form of progressive "data distillation". A deep learning model is like a sieve for data processing, made of a 
succession of increasingly refined data filters -- the "layers".

Here our network consists of a sequence of two `Dense` layers, which are densely-connected (also called "fully-connected") neural layers. 
The second (and last) layer is a 10-way "softmax" layer, which means it will return an array of 10 probability scores (summing to 1). Each 
score will be the probability that the current digit image belongs to one of our 10 digit classes.

To make our network ready for training, we need to pick three more things, as part of "compilation" step:

* A loss function: the is how the network will be able to measure how good a job it is doing on its training data, and thus how it will be 
able to steer itself in the right direction.
* An optimizer: this is the mechanism through which the network will update itself based on the data it sees and its loss function.
* Metrics to monitor during training and testing. Here we will only care about accuracy (the fraction of the images that were correctly 
classified).

The exact purpose of the loss function and the optimizer will be made clear throughout the next two chapters.

In [12]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])
# optimizer는 'rmsprop', loss함수는 'categorical_crossentropy', metrics은 'accuracy'러 씀을 선언


Before training, we will preprocess our data by reshaping it into the shape that the network expects, and scaling it so that all values are in 
the `[0, 1]` interval. Previously, our training images for instance were stored in an array of shape `(60000, 28, 28)` of type `uint8` with 
values in the `[0, 255]` interval. We transform it into a `float32` array of shape `(60000, 28 * 28)` with values between 0 and 1.

In [0]:
train_images = train_images.reshape((60000, 28 * 28))
# train_images 형태를 (60000,28*28) 로 reshape해줌
train_images = train_images.astype('float32') / 255
# train_images 값을 0에서 1사이값으로 변환

test_images = test_images.reshape((10000, 28 * 28))
# test_images 형태를 (10000,28*28) 로 reshape해줌
test_images = test_images.astype('float32') / 255
# test_images 값을 0에서 1사이값으로 변환

We also need to categorically encode the labels, a step which we explain in chapter 3:

In [0]:
from keras.utils import to_categorical
#keras.utils 에서 to_catogorical 모듈 활성화

train_labels = to_categorical(train_labels)
# train_labels를 범주형으로 변경
test_labels = to_categorical(test_labels)
# test_labels를 범주형으로 변경

We are now ready to train our network, which in Keras is done via a call to the `fit` method of the network: 
we "fit" the model to its training data.

In [15]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)
# batch sample은 128개씩 사용, epochs(한 데이터를 얼마나 돌리느냐)은 5로 지정해서 네트워크 적용

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Epoch 1/5





60000/60000 [==============================] - 5s 88us/step - loss: 0.2618 - acc: 0.9234
Epoch 2/5
60000/60000 [==============================] - 5s 76us/step - loss: 0.1074 - acc: 0.9681
Epoch 3/5
60000/60000 [==============================] - 5s 76us/step - loss: 0.0712 - acc: 0.9788
Epoch 4/5
60000/60000 [==============================] - 5s 76us/step - loss: 0.0529 - acc: 0.9844
Epoch 5/5
60000/60000 [==============================] - 5s 78us/step - loss: 0.0399 - acc: 0.9883


Two quantities are being displayed during training: the "loss" of the network over the training data, and the accuracy of the network over 
the training data.

We quickly reach an accuracy of 0.989 (i.e. 98.9%) on the training data. Now let's check that our model performs well on the test set too:

In [16]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
# 네트워크 평가해서 정확도와 loss값 산출

10000/10000 [==============================] - 1s 52us/step


0.07539410055298358

In [0]:
print('test_acc:', test_acc)

# 테스트 정확도 출력

test_acc: 0.9777



Our test set accuracy turns out to be 97.8% -- that's quite a bit lower than the training set accuracy. 
This gap between training accuracy and test accuracy is an example of "overfitting", 
the fact that machine learning models tend to perform worse on new data than on their training data. 
Overfitting will be a central topic in chapter 3.

This concludes our very first example -- you just saw how we could build and a train a neural network to classify handwritten digits, in 
less than 20 lines of Python code. In the next chapter, we will go in detail over every moving piece we just previewed, and clarify what is really 
going on behind the scenes. You will learn about "tensors", the data-storing objects going into the network, about tensor operations, which 
layers are made of, and about gradient descent, which allows our network to learn from its training examples.